### ISSUE TITLE
Sequence characteristic metadata

### ISSUE THREAD

**Issue by ahwagner** [ga4gh/vrs#431]
References to sequences via GA4GH `SQ.<digest>` identifiers has been a feature of VRS 1.0 since foundation. However, these digests (based on the RefGet spec) do not contain information about the sequence alphabet or the molecule circularity (circular / linear), making implementation of methods for sequence operations on different molecule types challenging. Some initial work is being planned in biocommons/biocommons.seqrepo#113 to implement and use these metadata for sequence operations. In VRS, we will need to consider how to capture these metadata as a separate structure, since a solution to this issue will not be addressed in the upcoming RefGet v2.

An initial implementation proposal is to create an annotation wrapper layer as a separate class, that could then be referenced by parent objects such as `SequenceLocation`. We have drafted such a class as an example, here: https://github.com/ga4gh/vrs/blob/c693ac3e8ac85d793a7fe73529504038aa71c815/schema/vrs-source.yaml#L422-L447.

Note this requires creating a separate type prefix (`SQR.`, for **S**e**q**uence **R**eference), as this class references a molecular sequence but is _distinct from_ the sequence as defined by RefGet (with reserved `SQ.` namespace). A drawback of this approach is an additional layer of indirection that implementations will need to support to get back to RefGet accessions.

Opening this for discussion and progress on this topic for VRS 2.0.

**Comment by andrewyatz** [comment]
I agree `residueAlphabet` is not an intrinsic part of the sequence and the advantages laid out here are important. I'm not sure if I think the designation of sequence type should be at the `SequenceReference` level or higher. `residueAlphabet` is good for disambiguating between dna/protein but doesn't help disambiguate between genomic or coding aside from protein (the 3 levels I am assuming variation should be recorded against). Would appreciate input on this point if it is seen as useful to know are we talking about a genomic variant or one against a coding sequence.

**Comment by rrfreimuth** [comment]
I think I understand the use cases and rationale underlying the proposal at Connect, but to be honest it feels like the model conflates a couple of concepts. This simplification may be acceptable. My primary concern is that the model supports the minimum level of semantic expressiveness that is required to meet all current (and hopefully near-future) use cases. If it does that, then it will also provide a degree of stability over time that will allow implementers to build off of it. Because these are such fundamental concepts at the core of so many use cases, changes at this level can have ripple effects through anything that uses it. Therefore, I think we should take a little time to ensure we get it as right as we can.

The discussion regarding inherent properties of sequence made clear (to me, anyway) that there are at least two concepts involved:

1. the sequence itself, which refers to the primary, linear arrangements of subunits[^1]
2. the representation of the sequence, which refers to how we instantiate the primary, linear arrangement

This distinction is reflected in this statement, which differentiates between the "meaning" and the "representation" of a sequence:

> Specifically, when a sequence can represent either a valid amino acid or a nucleic acid string, the meaning of the sequence (and variants on the sequence) is quite different. 

`M` vs. `Met`: Clearly different ASCII representations, but do they mean the same sequence?
`ATG` vs. `ATG`: Nucleotide start codon or tripeptide? Do they mean the same sequence?

In another example: The codon changed in EGFR T790M (2369C>T)[^2] is `ACG` to `ATG`, which can be expressed as `AYG` using IUPAC nucleotide ambiguity codes. All of those components (T, M, C, T, ACG, ATG, AYG) use valid nucleotide or protein alphabets. While most of those components clearly mean different things (based on string identity), one `T` represents threonine while another `T` represents thymine. Both are used in the context of describing a single variation, and both would have the same LiteralSequenceExpression[^3].

The fact that we have been wrestling with this issue suggests this distinction is important for at least some of our use cases, but because we are experimenting with both expanded and simplified models it might not be necessary in all implementations.

In another group[^4], our model supports the distinction between sequence "concept" and sequence representation (which includes a field for alphabet) because of the heterogeneity that exists in representation yet the need to support identity-based operations that determine if two sequences are semantically the same. A single sequence could be expressed using multiple representations, which permit domain-friendly conventions and which can be normalized to the same literal sequence expression to permit identity operations.

I think we need to determine whether it is important for us to be able to not only assess when two instances have identical textual serializations (the digest), but also if they mean the same thing. If we can come to conclusion on that, then the questions about what to include in the model (and how) might become more clear.



 [^1]: Note that the type of subunit may be a critical piece of information that defines the type of sequence, but this is not necessarily sufficient in all cases. Concepts such as the physical topology of the molecule (linear vs circular) and distinctions based on biological function (e.g., mRNA vs rRNA, or genomic DNA vs cDNA) represent additional concepts that are not captured by sequence or subunit type. Nonetheless, it may be convenient to model and represent these concepts as a single attribute unless use cases are presented that require additional levels of detail.

[^2]: [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/variation/16613/)

[^3]: Please correct me if my understanding is out of date.

[^4]: Sorry...
